File: apps/pi-worker/observability.test.mjs
import assert from 'node:assert/strict';
import { once } from 'node:events';
import { createServer } from 'node:http';
import test from 'node:test';
import { createWorkerObservability, loadObservabilityConfig, parseInternalParent } from './observability.mjs';
import { emptyUsage } from './usage.mjs';

function configuration(env = {}) {
  return { models: [{ id: 'catalog-model', provider: 'openai', protocol: 'responses' }], maxConcurrency: 4,
    observability: loadObservabilityConfig(env, 'pi', 'development') };
}
const event = { type: 'completed', text: 'private-output', observability: { usage: { ...emptyUsage(), status: 'reported' }, timing: { workerTotalMs: 120, setupMs: 10, providerMs: 80, providerTtftMs: 30, workerFirstDeltaMs: 40 } } };

test('management metrics are disabled by default and reject non-loopback configuration', async () => {
  const disabled = createWorkerObservability(configuration());
  assert.equal(configuration({ OTEL_TRACES_SAMPLER_ARG: '' }).observability.ratio, 1);
  assert.throws(() => configuration({ OTEL_EXPORTER_OTLP_ENDPOINT: 'https://api.openai.com/v1/traces' }), /internal/);
  assert.equal(disabled.server, undefined); await disabled.listen(); await disabled.close();
  for (const address of ['0.0.0.0:9102', 'localhost:9102', '192.168.1.2:9102', '127.0.0.1:99999']) assert.throws(() => configuration({ AWWO_METRICS_LISTEN_ADDR: address }), /loopback/);
  assert.throws(() => configuration({ AWWO_METRICS_ENABLED: 'yes' }), /true or false/);
  assert.throws(() => configuration({ AWWO_OTEL_ENABLED: 'true' }), /required/);
  assert.throws(() => configuration({ OTEL_EXPORTER_OTLP_ENDPOINT: 'https://key:secret@host/' }), /without credentials/);
  assert.throws(() => configuration({ OTEL_TRACES_SAMPLER_ARG: '2' }), /between/);
});

test('independent real management listener has bounded labels and terminal counters are idempotent', async t => {
  const config = configuration({ AWWO_METRICS_ENABLED: 'true', AWWO_METRICS_LISTEN_ADDR: '127.0.0.1:0' });
  const monitor = createWorkerObservability(config, () => 2); t.after(() => monitor.close()); await monitor.listen();
  const finish = monitor.begin(config.models[0], 'private-invalid-parent'); finish({ type: 'text_delta', delta: 'private-input' }); finish(event); finish(event);
  const base = `http://127.0.0.1:${monitor.server.address().port}`;
  const response = await fetch(`${base}/metrics`); assert.equal(response.status, 200);
  const metrics = await response.text();
  assert.match(metrics, /awwo_worker_invocations_total\{runtime="pi",model="catalog-model",outcome="completed",usage_status="reported"\} 1/);
  assert.match(metrics, /awwo_worker_provider_ttft_seconds_count.* 1/); assert.match(metrics, /awwo_worker_active 2/);
  assert.ok(!metrics.includes('private')); assert.ok(!metrics.includes('tenant'));
  assert.equal((await fetch(`${base}/health`)).status, 404); assert.equal((await fetch(`${base}/metrics?prompt=private`)).status, 404);
});

test('bounded OTLP export uses authenticated parent identity and excludes content and custom env attributes', async t => {
  const batches = [];
  const collector = createServer(async (request, response) => {
    let text = ''; for await (const chunk of request) text += chunk;
    batches.push({ path: request.url, body: JSON.parse(text) }); response.writeHead(200); response.end('{}');
  });
  collector.listen(0, '127.0.0.1'); await once(collector, 'listening');
  t.after(async () => { collector.closeAllConnections(); await new Promise(resolve => collector.close(resolve)); });
  const config = configuration({ AWWO_OTEL_ENABLED: 'true', OTEL_EXPORTER_OTLP_ENDPOINT: `http://127.0.0.1:${collector.address().port}`, OTEL_RESOURCE_ATTRIBUTES: 'tenant_id=private-tenant,prompt=private-prompt' });
  const monitor = createWorkerObservability(config); t.after(() => monitor.close());
  const trace = '1'.repeat(32), parent = '2'.repeat(16);
  const finish = monitor.begin(config.models[0], `00-${trace}-${parent}-01`); finish(event);
  await monitor.flush();
  assert.equal(batches.length, 1); assert.equal(batches[0].path, '/v1/traces');
  const spans = batches[0].body.resourceSpans[0].scopeSpans[0].spans;
  assert.equal(spans.length, 2); assert.equal(spans[0].name, 'worker.run'); assert.equal(spans[0].traceId, trace); assert.equal(spans[0].parentSpanId, parent);
  assert.equal(spans[1].parentSpanId, spans[0].spanId); assert.equal(spans[1].name, 'worker.provider.call');
  const serialized = JSON.stringify(batches); assert.ok(!serialized.includes('private')); assert.ok(!serialized.includes('tenant_id'));
  monitor.begin(config.models[0], `00-${trace}-${parent}-00`)(event); await monitor.flush(); assert.equal(batches.length, 1);
  assert.equal(parseInternalParent(`00-${'0'.repeat(32)}-${parent}-01`), null);
});

test('collector failure is isolated and the exporter queue stays bounded', async t => {
  const config = configuration({ AWWO_OTEL_ENABLED: 'true', OTEL_EXPORTER_OTLP_ENDPOINT: 'http://127.0.0.1:1' });
  const monitor = createWorkerObservability(config); t.after(() => monitor.close());
  for (let i = 0; i < 300; i++) monitor.begin(config.models[0])(event);
  assert.match(monitor.metrics(), /awwo_otel_dropped_spans_total 88/);
  await monitor.flush(); assert.match(monitor.metrics(), /awwo_otel_export_failures_total 1/);
  assert.match(monitor.metrics(), /usage_status="reported"\} 300/);
});

test('business listener never serves metrics and only an authenticated accepted run receives W3C parent', async t => {
  const { loadConfig } = await import('./config.mjs');
  const worker = await import('./server.mjs');
  const makeServer = worker.createPiServer ?? worker.createOpenAIAgentsServer;
  const token = 'private-auth-test-token-longer-than-32-characters';
  const config = loadConfig({
    AWWO_PI_PROVIDER: 'openai', AWWO_PI_BASE_URL: 'http://127.0.0.1:1/v1', AWWO_PI_TOKEN: token, AWWO_PI_MODEL: 'catalog-model', AWWO_PI_API_KEY: 'private-key',
    AWWO_OPENAI_AGENTS_TOKEN: token, AWWO_OPENAI_AGENTS_MODEL: 'catalog-model', AWWO_OPENAI_AGENTS_API_KEY: 'private-key',
    AWWO_METRICS_ENABLED: 'true', AWWO_METRICS_LISTEN_ADDR: '127.0.0.1:0',
  });
  let launched = 0;
  const app = makeServer(config, { startRun: async ({ onExit, onEvent }) => {
    launched++; onExit(); onEvent(event); return { cancel() {}, done: Promise.resolve() };
  } });
  t.after(() => app.close());
  await app.observability.listen(); app.server.listen(0, '127.0.0.1'); await once(app.server, 'listening');
  const url = `http://127.0.0.1:${app.server.address().port}`;
  assert.equal((await fetch(`${url}/metrics`, { headers: { authorization: `Bearer ${token}` } })).status, 404);
  const body = JSON.stringify({ runId: 'private-run', tenantId: 'private-tenant', sessionId: 'private-session', messages: [], prompt: 'private-prompt' });
  const headers = { 'content-type': 'application/json', traceparent: `00-${'1'.repeat(32)}-${'2'.repeat(16)}-01`, baggage: 'private-baggage' };
  assert.equal((await fetch(`${url}/internal/runs`, { method: 'POST', headers, body })).status, 401); assert.equal(launched, 0);
  assert.ok(!app.observability.metrics().includes('usage_status='));
  const response = await fetch(`${url}/internal/runs`, { method: 'POST', headers: { ...headers, authorization: `Bearer ${token}` }, body });
  assert.equal(response.status, 200); await response.text(); assert.equal(launched, 1);
  assert.match(app.observability.metrics(), /usage_status="reported"\} 1/);
  assert.ok(!app.observability.metrics().includes('private'));
});

File: backend/internal/app/observability_test.go
package app

import (
	"context"
	"encoding/base64"
	"io"
	"net/http"
	"net/http/httptest"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"testing"
	"time"

	"github.com/prometheus/client_golang/prometheus/promhttp"
	sdktrace "go.opentelemetry.io/otel/sdk/trace"
	"go.opentelemetry.io/otel/sdk/trace/tracetest"
	"go.opentelemetry.io/otel/trace"
	collectortrace "go.opentelemetry.io/proto/otlp/collector/trace/v1"
	"google.golang.org/protobuf/proto"
)

func metricText(t *testing.T, a *App) string {
	t.Helper()
	w := httptest.NewRecorder()
	promhttp.HandlerFor(a.obs.registry, promhttp.HandlerOpts{}).ServeHTTP(w, httptest.NewRequest("GET", "/metrics", nil))
	if w.Code != 200 {
		t.Fatalf("metrics status %d: %s", w.Code, w.Body.String())
	}
	return w.Body.String()
}

func TestObservabilityTraceHelpersWithoutInitialization(t *testing.T) {
	a := &App{}
	parent, cancel := context.WithCancel(context.WithValue(context.Background(), requestIDKey{}, "server-request"))
	cancel()
	for _, tc := range []struct {
		name  string
		start func(context.Context) (context.Context, func(string))
	}{
		{"internal", func(ctx context.Context) (context.Context, func(string)) {
			return a.startInternalTrace(ctx, "pi", "fixture")
		}},
		{"team", func(ctx context.Context) (context.Context, func(string)) {
			return a.startTeamTrace(ctx, "work", "pi", 1)
		}},
		{"graph", func(ctx context.Context) (context.Context, func(string)) { return a.startGraphNodeTrace(ctx, "pi") }},
		{"database", func(ctx context.Context) (context.Context, func(string)) { return a.startDBTrace(ctx, "run_read") }},
		{"ledger", func(ctx context.Context) (context.Context, func(string)) {
			return a.startLedgerTrace(ctx, "run_terminal_commit")
		}},
	} {
		t.Run(tc.name, func(t *testing.T) {
			ctx, end := tc.start(parent)
			if ctx.Err() != context.Canceled || RequestID(ctx) != "server-request" || trace.SpanFromContext(ctx).IsRecording() {
				t.Fatal("disabled tracing changed business context")
			}
			end("cancelled")
		})
	}
	ctx, end := a.startAsyncTrace(parent, "awwo.run.execute", "direct", "pi")
	if ctx.Err() != nil || RequestID(ctx) != "server-request" || trace.SpanFromContext(ctx).IsRecording() {
		t.Fatal("disabled async tracing changed background execution context")
	}
	end("completed")
	a.cfg.OTelEnabled = true
	if err := a.startTracing(context.Background()); err != nil || a.obs != nil {
		t.Fatal("tracing initialized for a caller without observability", err)
	}
}

func TestObservabilityManagementIsSeparateAndDisabledByDefault(t *testing.T) {
	a := New(nil, Config{Env: "development"})
	defer a.Close()
	if err := a.StartObservability(context.Background()); err != nil {
		t.Fatal(err)
	}
	if a.obs.listener != nil {
		t.Fatal("default mode opened metrics port")
	}
	w := httptest.NewRecorder()
	a.Handler().ServeHTTP(w, httptest.NewRequest("GET", "/metrics", nil))
	if w.Code != 404 || len(w.Header().Get("X-Request-ID")) != 32 {
		t.Fatalf("public metrics=%d request identity=%q", w.Code, w.Header().Get("X-Request-ID"))
	}
	b := New(nil, Config{Env: "development", MetricsEnabled: true, MetricsListenAddr: "127.0.0.1:0"})
	defer b.Close()
	if err := b.StartObservability(context.Background()); err != nil {
		t.Fatal(err)
	}
	for path, status := range map[string]int{"/metrics": 200, "/healthz": 204, "/api/v1/health": 404} {
		res, err := http.Get("http://" + b.obs.listener.Addr().String() + path)
		if err != nil {
			t.Fatal(err)
		}
		io.Copy(io.Discard, res.Body)
		res.Body.Close()
		if res.StatusCode != status {
			t.Fatalf("management %s: %d", path, res.StatusCode)
		}
		if res.Header.Get("Access-Control-Allow-Origin") != "" || res.Header.Get("Set-Cookie") != "" {
			t.Fatal("management endpoint acquired browser authorization surface")
		}
	}
	w = httptest.NewRecorder()
	b.Handler().ServeHTTP(w, httptest.NewRequest("GET", "/metrics", nil))
	if w.Code != 404 {
		t.Fatal("business metrics must stay 404 even when management is enabled")
	}
}

func TestObservabilityLoopbackAndConfigBoundaries(t *testing.T) {
	for _, address := range []string{"0.0.0.0:9101", "[::]:9101", "localhost:9101", "127.0.0.1.evil:9101", "192.168.1.5:9101", "127.0.0.1:65536", "127.0.0.1:-1", "127.0.0.1"} {
		if loopbackAddress(address) {
			t.Errorf("accepted unsafe address %s", address)
		}
	}
	for _, address := range []string{"127.0.0.1:9101", "[::1]:9101", "127.0.0.2:0"} {
		if !loopbackAddress(address) {
			t.Errorf("rejected loopback address %s", address)
		}
	}
	for _, c := range []Config{{Env: "production", OTelEnabled: true, OTelEndpoint: "https://provider.example/v1/traces"}, {Env: "staging", OTelEnabled: true, OTelEndpoint: "http://collector.internal:4318"}, {Env: "development", OTelEnabled: true, OTelEndpoint: "http://127.0.0.1:4318?key=secret"}, {Env: "development", OTelResourceAttributes: "tenant.id=secret"}, {UsageRetentionDays: 30}, {OTelSampleRatio: 1.1}, {OTelSampler: "always_on"}} {
		if c.validateObservability() == nil {
			t.Errorf("accepted unsafe config: env=%s", c.Env)
		}
	}
	for _, c := range []Config{{Env: "production", OTelEnabled: true, OTelEndpoint: "http://127.0.0.1:4318", OTelSampleRatio: .05}, {Env: "production", OTelEnabled: true, OTelEndpoint: "https://collector.internal:4318", OTelSampleRatio: .05}, {Env: "development", UsageRetentionDays: 180}} {
		if err := c.validateObservability(); err != nil {
			t.Fatal(err)
		}
	}
}

func TestObservabilityHTTPIdentityRoutesAndSSE(t *testing.T) {
	a := New(nil, Config{MetricsEnabled: true})
	defer a.Close()
	recorder := tracetest.NewSpanRecorder()
	provider := sdktrace.NewTracerProvider(sdktrace.WithSpanProcessor(recorder))
	a.obs.tracer = provider.Tracer("test")
	mux := http.NewServeMux()
	const pattern = "/api/v1/tenants/{tenantId}/runs/{id}/events"
	mux.HandleFunc("GET "+pattern, func(w http.ResponseWriter, r *http.Request) {
		if RequestID(r.Context()) == "spoofed-private-id" || RequestID(r.Context()) == "" {
			t.Error("client controls request id")
		}
		for _, key := range []string{"traceparent", "tracestate", "baggage", "X-Request-ID"} {
			if r.Header.Get(key) != "" {
				t.Errorf("untrusted %s reached handler", key)
			}
		}
		w.Header().Set("Content-Type", "text/event-stream")
		w.WriteHeader(200)
		w.(http.Flusher).Flush()
		time.Sleep(25 * time.Millisecond)
		io.WriteString(w, "data: private-response\n\n")
	})
	h := a.observeHTTP(mux, mux)
	req := httptest.NewRequest("GET", "/api/v1/tenants/private-tenant/runs/private-run/events?secret=query", nil)
	req.Header.Set("X-Request-ID", "spoofed-private-id")
	req.Header.Set("traceparent", "00-11111111111111111111111111111111-2222222222222222-01")
	req.Header.Set("baggage", "password=do-not-export")
	req.Header.Set("tracestate", "vendor=private-state")
	w := httptest.NewRecorder()
	h.ServeHTTP(w, req)
	if w.Code != 200 || !strings.Contains(w.Body.String(), "private-response") {
		t.Fatal("SSE response changed")
	}
	for i := 0; i < 3; i++ {
		h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", "/not-found/private-dynamic-id", nil))
	}
	metrics := metricText(t, a)
	for _, secret := range []string{"private-tenant", "private-run", "private-response", "private-dynamic-id", "secret=query", "spoofed-private-id", "password", "22222222"} {
		if strings.Contains(metrics, secret) {
			t.Errorf("metrics leaked %s", secret)
		}
	}
	if !strings.Contains(metrics, `route="`+pattern+`"`) || !strings.Contains(metrics, `stream="true"`) || !strings.Contains(metrics, "awwo_sse_handshake_duration_seconds_count") || !strings.Contains(metrics, `route="unmatched"`) {
		t.Fatal("route or SSE metrics missing")
	}
	spans := recorder.Ended()
	if len(spans) != 4 {
		t.Fatalf("spans=%d", len(spans))
	}
	if spans[0].Parent().IsValid() || spans[0].SpanContext().TraceID().String() == "11111111111111111111111111111111" {
		t.Fatal("public W3C parent was trusted")
	}
	families, _ := a.obs.registry.Gather()
	var duration, handshake float64
	for _, f := range families {
		if f.GetName() == "awwo_sse_handshake_duration_seconds" {
			handshake = f.Metric[0].Histogram.GetSampleSum()
		}
		if f.GetName() == "awwo_http_request_duration_seconds" {
			for _, m := range f.Metric {
				for _, label := range m.Label {
					if label.GetName() == "stream" && label.GetValue() == "true" {
						duration = m.Histogram.GetSampleSum()
					}
				}
			}
		}
	}
	if duration < .02 || handshake >= duration {
		t.Fatalf("SSE timings do not separate headers and lifetime: %f/%f", handshake, duration)
	}
}

func TestObservabilitySSERejectedHandshakeAndNoContent(t *testing.T) {
	a := New(nil, Config{MetricsEnabled: true})
	defer a.Close()
	mux := http.NewServeMux()
	mux.HandleFunc("GET /api/v1/tenants/{tenantId}/runs/{id}/events", func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(401) })
	h := a.observeHTTP(mux, mux)
	h.ServeHTTP(httptest.NewRecorder(), httptest.NewRequest("GET", "/api/v1/tenants/t/runs/r/events", nil))
	metrics := metricText(t, a)
	if !strings.Contains(metrics, `status_class="4xx",stream="true"`) || !strings.Contains(metrics, "awwo_sse_handshake_duration_seconds_count") {
		t.Fatal("failed SSE lost registered stream identity")
	}
	mux.HandleFunc("GET /empty", func(w http.ResponseWriter, r *http.Request) {})
	w := httptest.NewRecorder()
	h.ServeHTTP(w, httptest.NewRequest("GET", "/empty", nil))
	if w.Code != 200 {
		t.Fatal("implicit header commit changed")
	}
}

func TestObservabilityInvocationUsesBoundedCatalogAndNullUsage(t *testing.T) {
	a := New(nil, Config{MetricsEnabled: true})
	defer a.Close()
	a.registerTelemetryModels("pi", []piModel{{ID: "safe-model"}})
	n := int64(12)
	zero := int64(0)
	a.observeInvocation("pi", "safe-model", "completed", "reported", "", map[string]*int64{"input": &n, "output": &zero}, nil, nil, nil)
	a.observeInvocation("pi", "private-unregistered-model", "failed", "unknown", "raw-sensitive-error", nil, nil, nil, nil)
	metrics := metricText(t, a)
	if strings.Contains(metrics, "private-unregistered-model") || strings.Contains(metrics, "raw-sensitive-error") {
		t.Fatal("unbounded labels leaked")
	}
	if !strings.Contains(metrics, `direction="input",model="safe-model",runtime="pi"} 12`) || !strings.Contains(metrics, `direction="output",model="safe-model",runtime="pi"} 0`) {
		t.Fatal("explicit usage zero or input missing")
	}
	if strings.Contains(metrics, "awwo_model_time_to_first_token_seconds_count") || strings.Contains(metrics, "awwo_model_estimated_cost_usd_total{") {
		t.Fatal("unknown timing/cost became zero")
	}
}

func TestObservabilityAsyncLinkAndInternalPropagation(t *testing.T) {
	a := New(nil, Config{PIURL: "http://127.0.0.1:8097", PIToken: strings.Repeat("x", 32)})
	defer a.Close()
	recorder := tracetest.NewSpanRecorder()
	provider := sdktrace.NewTracerProvider(sdktrace.WithSpanProcessor(recorder))
	a.obs.tracer = provider.Tracer("test")
	parent, cancel := context.WithCancel(context.WithValue(context.Background(), requestIDKey{}, "server-request"))
	parent, span := a.obs.tracer.Start(parent, "http.server")
	parentSC := span.SpanContext()
	span.End()
	cancel()
	ctx, end := a.startAsyncTrace(parent, "awwo.run.execute", "direct", "pi")
	if ctx.Err() != nil {
		t.Fatal("async task inherited completed request cancellation")
	}
	admission, finish := a.startInternalTrace(ctx, "pi", "secret-model")
	req, _ := http.NewRequestWithContext(admission, "POST", a.cfg.PIURL+"/internal/runs", nil)
	req.Header.Set("Authorization", "Bearer "+a.cfg.PIToken)
	req.Header.Set("baggage", "sensitive")
	req.Header.Set("tracestate", "sensitive")
	a.injectWorkerTrace(req, "pi")
	if req.Header.Get("traceparent") == "" || req.Header.Get("X-Request-ID") != "server-request" || req.Header.Get("baggage") != "" || req.Header.Get("tracestate") != "" {
		t.Fatal("internal trace boundary broken")
	}
	external, _ := http.NewRequestWithContext(admission, "POST", "https://provider.example/internal/runs", nil)
	external.Header.Set("traceparent", req.Header.Get("traceparent"))
	a.injectWorkerTrace(external, "pi")
	if external.Header.Get("traceparent") != "" {
		t.Fatal("trace crossed the configured internal endpoint")
	}
	finish("accepted")
	end("completed")
	spans := recorder.Ended()
	run := spans[len(spans)-1]
	if run.Parent().IsValid() || len(run.Links()) != 1 || run.Links()[0].SpanContext.TraceID() != parentSC.TraceID() {
		t.Fatal("async trace is not a new root linked to admission")
	}
	if run.SpanContext().TraceID() == parentSC.TraceID() {
		t.Fatal("async task reused HTTP trace")
	}
}

func TestObservabilityActualOTLPExportIsContentFree(t *testing.T) {
	var mu sync.Mutex
	var received []*collectortrace.ExportTraceServiceRequest
	collector := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		bytes, _ := io.ReadAll(io.LimitReader(r.Body, 2<<20))
		var message collectortrace.ExportTraceServiceRequest
		if err := proto.Unmarshal(bytes, &message); err != nil {
			t.Error(err)
		}
		if r.URL.Path != "/v1/traces" {
			t.Errorf("wrong OTLP path %s", r.URL.Path)
		}
		mu.Lock()
		received = append(received, &message)
		mu.Unlock()
		w.Header().Set("Content-Type", "application/x-protobuf")
		w.WriteHeader(200)
	}))
	defer collector.Close()
	a := New(nil, Config{Env: "development", OTelEnabled: true, OTelEndpoint: collector.URL, OTelSampleRatio: 1})
	if err := a.StartObservability(context.Background()); err != nil {
		t.Fatal(err)
	}
	mux := http.NewServeMux()
	mux.HandleFunc("POST /test/{id}", func(w http.ResponseWriter, r *http.Request) {
		io.Copy(io.Discard, r.Body)
		io.WriteString(w, "private-output-marker")
	})
	r := httptest.NewRequest("POST", "/test/private-business-id?secret=private-query", strings.NewReader("private-prompt-marker"))
	r.Header.Set("Authorization", "Bearer private-provider-key")
	r.Header.Set("Cookie", "private-cookie")
	r.Header.Set("traceparent", "00-11111111111111111111111111111111-2222222222222222-01")
	a.observeHTTP(mux, mux).ServeHTTP(httptest.NewRecorder(), r)
	a.Close()
	mu.Lock()
	defer mu.Unlock()
	if len(received) == 0 {
		t.Fatal("collector received no actual OTLP request")
	}
	for _, message := range received {
		b, _ := proto.Marshal(message)
		for _, secret := range []string{"private-prompt", "private-output", "private-business", "private-query", "private-provider", "private-cookie", "1111111111111111"} {
			if strings.Contains(string(b), secret) {
				t.Fatalf("OTLP leaked %s", secret)
			}
		}
	}
}

func TestObservabilityExporterFailureDoesNotBlockRequests(t *testing.T) {
	collector := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(503)
		io.WriteString(w, "private-collector-error")
	}))
	defer collector.Close()
	a := New(nil, Config{Env: "development", MetricsEnabled: true, MetricsListenAddr: "127.0.0.1:0", OTelEnabled: true, OTelEndpoint: collector.URL, OTelSampleRatio: 1})
	if err := a.StartObservability(context.Background()); err != nil {
		t.Fatal(err)
	}
	mux := http.NewServeMux()
	mux.HandleFunc("GET /ok", func(w http.ResponseWriter, r *http.Request) { w.WriteHeader(204) })
	h := a.observeHTTP(mux, mux)
	start := time.Now()
	for i := 0; i < 1200; i++ {
		w := httptest.NewRecorder()
		h.ServeHTTP(w, httptest.NewRequest("GET", "/ok", nil))
		if w.Code != 204 {
			t.Fatal("collector failure changed request")
		}
	}
	if time.Since(start) > 2*time.Second {
		t.Fatal("request path waited for bounded exporter queue")
	}
	a.Close()
	metrics := metricText(t, a)
	if !strings.Contains(metrics, "awwo_telemetry_export_errors_total 1") && !strings.Contains(metrics, "awwo_telemetry_export_errors_total ") {
		t.Fatal("export failure was not counted")
	}
	if strings.Contains(metrics, "private-collector-error") {
		t.Fatal("raw collector error leaked")
	}
}

func TestObservabilityTraceReferenceUsesIndependentKey(t *testing.T) {
	dir := t.TempDir()
	path := filepath.Join(dir, "trace-key")
	key := []byte(strings.Repeat("z", 32))
	if err := os.WriteFile(path, []byte(base64.RawURLEncoding.EncodeToString(key)), 0600); err != nil {
		t.Fatal(err)
	}
	parsed, err := readTraceKey(path, "v1")
	if err != nil {
		t.Fatal(err)
	}
	a := New(nil, Config{TraceRefKey: parsed, TraceRefVersion: "v1"})
	defer a.Close()
	attrs := a.traceReference("run", "secret-id")
	if len(attrs) != 2 || len(attrs[0].Value.AsString()) != 32 || strings.Contains(attrs[0].Value.AsString(), "secret-id") {
		t.Fatal("HMAC trace ref invalid")
	}
	if len(a.traceReference("tenant", "secret-id")) != 0 {
		t.Fatal("unapproved ref type accepted")
	}
	if _, err = readTraceKey(path, ""); err == nil {
		t.Fatal("unversioned key accepted")
	}
	if trace.SpanContextFromContext(context.Background()).IsValid() {
		t.Fatal("unexpected root")
	}
}

func TestObservabilityWorkerRedirectDoesNotForwardCredentials(t *testing.T) {
	calls := 0
	other := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { calls++ }))
	defer other.Close()
	source := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) { http.Redirect(w, r, other.URL, 302) }))
	defer source.Close()
	a := New(nil, Config{RunTimeout: time.Second})
	defer a.Close()
	req, _ := http.NewRequest("GET", source.URL, nil)
	req.Header.Set("Authorization", "Bearer private-token")
	res, err := a.client.Do(req)
	if err != nil {
		t.Fatal(err)
	}
	res.Body.Close()
	if res.StatusCode != 302 || calls != 0 {
		t.Fatal("internal redirects followed")
	}
}
